fvm/evm/handler/handler.go

```diff
@@ -0,0 +1,414 @@
+	blockchain       types.BlockChain
+	backend          types.Backend
+	emulator         types.Emulator
+	addressAllocator types.AddressAllocator
+	newBlockDraft    *types.Block
+}
+
+var _ types.ContractHandler = &ContractHandler{}
+
+func NewContractHandler(
+	blockchain types.BlockChain,
+	addressAllocator types.AddressAllocator,
+	backend types.Backend,
+	emulator types.Emulator,
+) *ContractHandler {
+	return &ContractHandler{
+		blockchain:       blockchain,
+		addressAllocator: addressAllocator,
+		backend:          backend,
+		emulator:         emulator,
+	}
+}
+
+// AllocateAddress allocates an address to be used by the bridged accounts
+func (h *ContractHandler) AllocateAddress() types.Address {
+	target, err := h.addressAllocator.AllocateAddress()
+	handleError(err)
+	return target
+}
+
+// AccountByAddress returns the account for the given address,
+// if isAuthorized is set, account is controlled by the FVM (bridged accounts)
+func (h *ContractHandler) AccountByAddress(addr types.Address, isAuthorized bool) types.Account {
+	return newAccount(h, addr, isAuthorized)
+}
+
+// LastExecutedBlock returns the last executed block
+func (h ContractHandler) LastExecutedBlock() *types.Block {
+	block, err := h.blockchain.LatestBlock()
+	handleError(err)
+	return block
+}
+
+// Run runs an rlpencoded evm transaction and
+// collects the gas fees and pay it to the coinbase address provided.
+func (h ContractHandler) Run(rlpEncodedTx []byte, coinbase types.Address) {
+
+	// step 1 - transaction decoding
+	err := h.backend.MeterComputation(environment.ComputationKindRLPDecoding, uint(len(rlpEncodedTx)))
+	handleError(err)
+
+	tx := gethTypes.Transaction{}
+	err = tx.DecodeRLP(
+		rlp.NewStream(
+			bytes.NewReader(rlpEncodedTx),
+			uint64(len(rlpEncodedTx))))
+	handleError(err)
+
+	// step 2 - run transaction
+	h.checkGasLimit(types.GasLimit(tx.Gas()))
+
+	ctx := h.getBlockContext()
+	ctx.GasFeeCollector = coinbase
+	blk, err := h.emulator.NewBlockView(ctx)
+	handleError(err)
+
+	res, err := blk.RunTransaction(&tx)
+	h.meterGasUsage(res)
+	handleError(err)
+
+	// step 3 - process block and emit events
+	txHash := tx.Hash()
+	h.appendTxHashToBlockDraft(txHash)
+
+	h.emitEvent(types.NewTransactionExecutedEvent(
+		h.getBlockDraftHeight(),
+		rlpEncodedTx,
+		txHash,
+		res,
+	))
+	h.updateBlockDraftStateRoot(res.StateRootHash)
+	h.commitBlockDraft()
+}
+
+// this method has been separated out to make NewContractHandler lighter
+// and lazy load these operations if needed
+func (h *ContractHandler) setupNewBlockDraft() {
+	lastExecutedBlock, err := h.blockchain.LatestBlock()
+	handleError(err)
+
+	parentHash, err := lastExecutedBlock.Hash()
+	if err != nil {
+		// this is a fatal error
+		panic(err)
+	}
+	h.newBlockDraft = &types.Block{
+		Height:            lastExecutedBlock.Height + 1,
+		ParentBlockHash:   parentHash,
+		TotalSupply:       lastExecutedBlock.TotalSupply,
+		TransactionHashes: make([]gethCommon.Hash, 0),
+	}
+}
```

Comment:
wouldn't this make sense to be part of the blockchain interface and implemented as part of the blocks? the block model could also implement all the methods below to update the defined below.

Reply:
also this method could be the one to check:
```
if h.newBlockDraft != nil {
    return h.newBlockDraft
}

// otherwise lazy load it
```

Reply:
refactored at [3677c74](https://github.com/onflow/flow-go/pull/4864/commits/3677c741818a8ede653ab52d7b022faa6d8f7d8b)